pkg/agent/plugin/nodeattestor/tpmdevid/tpmutil/signingkey.go

```diff
@@ -0,0 +1,180 @@
+	}
+
+	return &SigningKey{
+		Handle:     keyHandle,
+		sigHashAlg: sigHashAlg,
+		rw:         rw,
+		log:        log,
+	}, nil
+}
+
+// Close removes the key from the TPM
+func (k *SigningKey) Close() error {
+	return tpm2.FlushContext(k.rw, k.Handle)
+}
+
+// Sign request the TPM to sign the given data using this key
+func (k *SigningKey) Sign(data []byte) ([]byte, error) {
+	digest, token, err := tpm2.Hash(k.rw, k.sigHashAlg, data, tpm2.HandlePlatform)
+	if err != nil {
+		return nil, fmt.Errorf("tpm2.Hash failed: %w", err)
+	}
+
+	var sig *tpm2.Signature
+	for i := 0; i <= maxAttempts; i++ {
+		sig, err = tpm2.Sign(k.rw, k.Handle, "", digest, token, nil)
+		switch {
+		case err == nil:
+			break
+
+		case isRetry(err):
+			if i == maxAttempts {
+				return nil, fmt.Errorf("max attempts reached: %w", err)
+			}
+
+			k.log.Debug(fmt.Sprintf("TPM was not able to start the command 'Sign'. Retrying: attempt (%d/%d)", i, maxAttempts))
+			time.Sleep(time.Millisecond * 500)
+			continue
+
+		default:
+			return nil, fmt.Errorf("tpm2.Sign failed: %w", err)
+		}
+	}
+
+	if sig.RSA != nil {
+		return sig.RSA.Signature, nil
+	}
+
+	if sig.ECC != nil {
+		var b cryptobyte.Builder
+		b.AddASN1(asn1.SEQUENCE, func(b *cryptobyte.Builder) {
+			b.AddASN1BigInt(sig.ECC.R)
+			b.AddASN1BigInt(sig.ECC.S)
+		})
+
+		return b.Bytes()
+	}
+
+	return nil, fmt.Errorf("bad tpm2.Signature")
+}
+
+// Certify calls tpm2.Certify using the current key as signer and the provided
+// object as object handle.
+func (k *SigningKey) Certify(object tpmutil.Handle) ([]byte, []byte, error) {
+	// For some reason 'tpm2.Certify()' sometimes fails the first attempt and asks for retry.
+	// So, we retry some times in case of getting the RCRetry error.
+	// It seems that this issue has been reported: https://github.com/google/go-tpm/issues/59
+	var certifiedDevID []byte
+	var certificationSignature []byte
+	var err error
+	for i := 0; i <= maxAttempts; i++ {
+		certifiedDevID, certificationSignature, err = tpm2.Certify(k.rw, "", "", object, k.Handle, nil)
+		switch {
+		case err == nil:
+			return certifiedDevID, certificationSignature, nil
+
+		case isRetry(err):
+			if i == maxAttempts {
+				return nil, nil, fmt.Errorf("max attempts reached: %w", err)
+			}
+
+			k.log.Debug(fmt.Sprintf("TPM was not able to start the command 'Certify'. Retrying: attempt (%d/%d)", i, maxAttempts))
+			time.Sleep(time.Millisecond * 500)
+
+		default:
+			break
+		}
+	}
+
+	return nil, nil, fmt.Errorf("certify failed: %w", err)
+}
+
+// SRKTemplateHighRSA returns the default high range SRK template (called H-1 in the specification).
+// https://trustedcomputinggroup.org/resource/tcg-ek-credential-profile-for-tpm-family-2-0/
+func SRKTemplateHighRSA() tpm2.Public {
+	// The tpm2tools library does not have a function to build the high range template
+	// so we build it based on the previous template.
+	template := tpm2tools.SRKTemplateRSA()
+	template.RSAParameters.ModulusRaw = []byte{}
```

Comment:
Yes, according to [TCG TPM v2.0 Provisioning Guidance (section 7.5.1)](https://trustedcomputinggroup.org/wp-content/uploads/TCG-TPM-v2.0-Provisioning-Guidance-Published-v1r1.pdf#page=26) the SRK template is based on EK template plus some changes in the ObjectAttributes and authPolicy fields.

Latest version of [TCG EK Credential Profile (section B.4.4)](https://trustedcomputinggroup.org/wp-content/uploads/TCG_IWG_EKCredentialProfile_v2p3_r2_pub.pdf#page=41) defines the H-1 RSA 2048 EK template with an empty buffer for the "unique" field.

I have been told that TCG discussed the use of the unique field and decided that it was better to have an empty buffer rather than a well-known buffer (filled with 0s for example). I don't know what is the exact reason, I think it may be related to TPM memory optimization.